tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,447 @@
+  if (root->opcode() != HloOpcode::kTuple) {
+    return 1;
+  } else {
+    return ShapeUtil::TupleElementCount(root->shape());
+  }
+}
+
+class HorizontalFusionImpl {
+ public:
+  HorizontalFusionImpl(HloComputation* computation)
+      : computation_(computation) {}
+
+  ~HorizontalFusionImpl() {}
+
+  StatusOr<bool> Run();
+
+ private:
+  Status Fuse(absl::Span<HloInstruction*> fused_fusion_insns);
+
+  Status CreateFusedComputation(
+      absl::Span<HloInstruction*> fused_fusion_insns,
+      std::unique_ptr<HloComputation>* uniq_computation,
+      std::vector<HloInstruction*>* bound_operands);
+
+  // FusionCandidates collects profitable candidates for a given consumer
+  // instruction. GetNextSpanOfFusions() can then be iteratively invoked to
+  // acquire the next set of fusion candidates based on some heuristics.
+  class FusionCandidates {
+   public:
+    FusionCandidates(HloInstruction* consumer) : fusion_insns_(), pos_(0) {
+      Initialize(consumer);
+    }
+
+    // Get a span of fusions to be fused.
+    absl::Span<HloInstruction*> GetNextSpanOfFusions();
+
+   private:
+    void Initialize(HloInstruction*);
+
+    std::vector<HloInstruction*> fusion_insns_;
+    // pos_ points to the start position of next span.
+    size_t pos_;
+  };
+
+  HloComputation* computation_;
+};  // HorizontalFusionImpl
+
+void HorizontalFusionImpl::FusionCandidates::Initialize(
+    HloInstruction* consumer) {
+  // First, find out all fusion instructions. We will filter out
+  // unsupported/non-profitable cases below.
+  absl::flat_hash_set<HloInstruction*> fusion_insns;
+  for (auto opnd : consumer->operands()) {
+    auto predecessor = opnd->LatestNonGteAncestor();
+    if (predecessor->opcode() == HloOpcode::kFusion) {
+      fusion_insns.insert(predecessor);
+    }
+  }
+
+
+  auto is_supported_fusion = [&](const HloInstruction& insn) -> bool {
+    // Support only kLoop fusion.
+    if (!insn.IsLoopFusion()) {
+      return false;
+    }
+
+    return true;
+  };
+
+  std::function<bool(const HloInstruction&)> is_consumer_the_only_user =
+      [&](const HloInstruction& insn) {
+        return absl::c_all_of(insn.users(), [&](const HloInstruction* user) {
+          if (user->opcode() == HloOpcode::kGetTupleElement) {
+            // Skip GTE.
+            return is_consumer_the_only_user(*user);
+          } else if (user == consumer) {
+            // user is consumer.
+            return true;
+          } else if (user == user->parent()->root_instruction()) {
+            // Consumed by ROOT is always fine, since it is impossible to create
+            // cycles through ROOT.
+            return true;
+          } else {
+            return false;
+          }
+        });
+      };
+
+  auto is_profitable = [](const HloInstruction& insn) -> bool {
+    // Do not fuse computation with too large shapes or too many instructions,
+    // because it may not be easily profitable.
+    constexpr int64 kShapeThreshold = 128 * 2048;
```

Comment:
Would it make sense to define a threshold on the number of bytes rather than the number of elements?